gen/fault-domain-detect/cloud.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/sh
+
+AWS_URL="http://169.254.169.254/latest/dynamic/instance-identity/document"
+
+AZURE_REGION_URL="http://169.254.169.254/metadata/instance/compute/location?api-version=2017-08-01&format=text"
+AZURE_FD_URL="http://169.254.169.254/metadata/instance/compute/platformFaultDomain?api-version=2017-04-02&format=text"
+
+GCP_METADATA_URL="http://metadata.google.internal/computeMetadata/v1/instance/zone"
+
+
+function aws() {
+    METADATA="$(curl -f -m3 $AWS_URL 2>/dev/null)"
+    rc=$?
+    if [ $rc -ne 0 ]; then
+        echo "unable to fetch aws region/zone. URL $AWS_URL. Ret code $rc" >&2
+        exit 1
+    fi
+    REGION=$(echo $METADATA | grep -Po "\"region\"\s+:\s+\"(.*?)\"" | cut -f2 -d: | tr -d ' \"')
+    ZONE=$(echo $METADATA | grep -Po "\"availabilityZone\"\s+:\s+\"(.*?)\"" | cut -f2 -d: | tr -d ' \"')
+    echo "{\"fault_domain\":{\"region\":{\"name\": \"aws/$REGION\"},\"zone\":{\"name\": \"aws/$ZONE\"}}}"
+    return 0
+}
+
+function azure() {
+    REGION=$(curl -f -m3 -H Metadata:true "$AZURE_REGION_URL" 2>/dev/null)
+    rc=$?
+    if [ $rc -ne 0 ]; then
+      echo "unable to fetch azure region. URL $AZURE_REGION_URL. Ret code $rc" >&2
+      exit 1
+    fi
+
+    FAULT_DOMAIN=$(curl -f -m3 -H Metadata:true "$AZURE_FD_URL" 2>/dev/null)
+    rc=$?
+    if [ $rc -ne 0 ]; then
+      echo "unable to fetch azure fault domain. URL $AZURE_FD_URL. Ret code $rc" >&2
+      exit 1
+    fi
+
+    echo "{\"fault_domain\":{\"region\":{\"name\": \"azure/$REGION\"},\"zone\":{\"name\": \"azure/$FAULT_DOMAIN\"}}}"
+    return 0
+}
+
+function gcp() {
+    BODY=$(curl -f -m3 -H "Metadata-Flavor: Google" "$GCP_METADATA_URL" 2>/dev/null)
+    rc=$?
+    if [ $rc -ne 0 ]; then
+      echo "unable to fetch gcp metadata. URL $GCP_METADATA_URL. Ret code $rc" >&2
+      exit 1
+    fi
+
+    ZONE=$(echo "$BODY" | sed 's@^projects/.*/zones/\(.*\)$@\1@')
+    REGION=$(echo "$ZONE" | sed 's@\(.*-.*\)-.*@\1@')
+
+    echo "{\"fault_domain\":{\"region\":{\"name\": \"gcp/$REGION\"},\"zone\":{\"name\": \"gcp/$ZONE\"}}}"
+    return 0
+}
+
+function main() {
+    if [ $# -eq 1 ]; then
```

Comment:
I'm trying to understand how this script will work in a hybrid cloud scenario? Who actually passes the `--aws`, `--azure`, or `--gcp` flag to this? The goal of this is to make sure that we have one script that on-prem users can place under their `/genconf` folder so that they can both (1) add their own custom on-prem logic to it, and (2) expand their cluster into any supported cloud provider without needing to reconfigure.